operators/pkg/exmagnt/instance.go

```diff
@@ -0,0 +1,270 @@
+)
+
+// ServeHTTP is the Instance handler for the exmagnt.
+func (i EAInstanceHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	klog.Infof("Request received from %s: %s@%s %s", r.RemoteAddr, r.Method, r.URL.Path, r.URL.RawQuery)
+
+	switch r.Method {
+	case http.MethodGet:
+		if GetInstanceIDFromRequest(r) == "" {
+			HandleGetAll(w, r)
+		} else {
+			i.HandleGet(w, r)
+		}
+	case http.MethodPut:
+		i.HandlePut(w, r)
+	case http.MethodDelete:
+		i.HandleDelete(w, r)
+	default:
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		fmt.Fprintf(w, "Method not allowed")
+		return
+	}
+}
+
+// HandleGet handles the GET request for the exmagnt.
+func (i *EAInstanceHandler) HandleGet(w http.ResponseWriter, r *http.Request) {
+	inst := EmptyInstanceFromRequest(r)
+
+	if err := Client.Get(r.Context(), forge.NamespacedName(inst), inst); err != nil {
+		if errors.IsNotFound(err) {
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprint(w, "Not found")
+			klog.Errorf("GET of not found Instance: %s", inst.Name)
+			return
+		}
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, "Error retrieving Instance")
+		klog.Errorf("Error retrieving Instance %s: %v", inst.Name, err)
+		return
+	}
+
+	if inst.Status.Phase == clv1alpha2.EnvironmentPhaseReady {
+		klog.Infof("Redirecting %v : %s", inst.Name, inst.Status.URL)
+		http.Redirect(w, r, inst.Status.URL, http.StatusFound)
+		return
+	}
+
+	if inst.Status.Phase == clv1alpha2.EnvironmentPhaseFailed || inst.Status.Phase == clv1alpha2.EnvironmentPhaseCreationLoopBackoff {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, "Something went wrong. Please retry later")
+		klog.Errorf("Instance %s INVALID PHSE: %v", inst.Name, inst.Status.Phase)
+		return
+	}
+
+	klog.Infof("Instance %s (phase: %v): sending starting-up page", inst.Name, inst.Status.Phase)
+
+	w.Header().Add("refresh", "5")
+	w.WriteHeader(http.StatusCreated)
+	fmt.Fprint(w, httpPageStartingUp)
+}
+
+// HandlePut handles the PUT request for a EAInstance api call.
+func (i *EAInstanceHandler) HandlePut(w http.ResponseWriter, r *http.Request) {
+	if !Options.CheckAllowedIP(r) {
+		klog.Errorf("Request from unauthorized IP: %v", r.RemoteAddr)
+		w.WriteHeader(http.StatusForbidden)
+		fmt.Fprint(w, "Forbidden")
+		return
+	}
+
+	// get Instance from the request
+	eaInstance, err := EAInstanceFromRequest(r)
+	if err != nil {
+		klog.Errorf("Error parsing request: %v", err)
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprint(w, "Bad request")
+		return
+	}
+
+	inst := EmptyInstanceFromRequest(r)
+	instSpec := InstanceSpecFromEAInstance(&eaInstance)
+
+	op, err := ctrl.CreateOrUpdate(r.Context(), Client, inst, func() error {
+		inst.Spec = instSpec
+		return nil
+	})
+
+	if err != nil {
+		klog.Errorf("Instance %s cannot be %s: %v", inst.Name, op, err)
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "Instance %s cannot be %s", inst.Name, op)
+		return
+	}
+
+	w.Header().Add("Content-Type", "text/plain")
```

Comment:
It is probably better to always stick to json (BTW, might it be worth returning the created instance information?)